File: packages/core/src/parser/__testdata__/sample.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;

namespace Synapse.Core
{
    public enum Status { Active, Idle, Stopped, Error }

    public record AgentConfig(string Model = "claude-opus-4-6", int MaxTokens = 4096, double Temperature = 0.7);

    public interface IAgent
    {
        Task<string> ProcessAsync(string message);
        IReadOnlyList<string> GetTools();
        Status Status { get; }
    }

    public interface IConfigurable
    {
        void Configure(AgentConfig config);
    }

    public struct Point
    {
        public double X { get; init; }
        public double Y { get; init; }
    }

    public abstract class BaseAgent : IAgent
    {
        protected readonly ILogger _logger;
        public const int MaxRetries = 3;

        protected BaseAgent(ILogger logger)
        {
            _logger = logger;
        }

        public abstract Task<string> ProcessAsync(string message);
        public abstract IReadOnlyList<string> GetTools();
        public Status Status { get; protected set; } = Status.Idle;
    }

    public class SynapseAgent : BaseAgent, IConfigurable
    {
        public string Name { get; }
        private AgentConfig _config;

        public SynapseAgent(string name, AgentConfig config, ILogger<SynapseAgent> logger)
            : base(logger)
        {
            Name = name;
            _config = config;
        }

        public override async Task<string> ProcessAsync(string message)
        {
            Status = Status.Active;
            _logger.LogInformation("Processing: {Message}", message);
            await Task.Delay(10);
            Status = Status.Idle;
            return $"Response to: {message}";
        }

        public override IReadOnlyList<string> GetTools()
            => new[] { "search", "read", "write" };

        public void Configure(AgentConfig config)
        {
            _config = config;
        }

        private bool Validate(string input) => !string.IsNullOrWhiteSpace(input);

        public static SynapseAgent Create(string name)
            => new(name, new AgentConfig(), null!);
    }

    // Verschachtelung fuer den findParentType-Fall: Entry steht DIREKT am Anfang
    // von Registry, ohne schliessende Klammer dazwischen. Bis Parser-Version 3
    // lieferte der Eltern-Typ hier die AEUSSERE Deklaration (Registry) statt Entry.
    // Describe/DescribeTwice und Lookup/Count decken zusaetzlich den haeufigeren
    // Fall ab: ab der ZWEITEN Methode ging der Eltern-Typ ganz verloren.
    public class Registry
    {
        public class Entry
        {
            public string Key;

            public string Describe() => Key;

            public string DescribeTwice() => Key + Key;
        }

        public Entry Lookup(string key) => null!;

        public int Count() => 0;
    }

    public delegate void StatusChanged(Status oldStatus, Status newStatus);

    // TODO: add health check endpoint
    // FIXME: logger null reference in Create method
}

File: packages/core/src/parser/__testdata__/sample.scala
package com.synapse.core

import scala.concurrent.{Future, ExecutionContext}
import scala.util.{Try, Success, Failure}

object Constants {
  val MaxRetries: Int = 3
  val DefaultModel: String = "claude-opus-4-6"
}

sealed trait Status
case object Active extends Status
case object Idle extends Status
case object Stopped extends Status
case class Error(message: String) extends Status

case class AgentConfig(
  model: String = Constants.DefaultModel,
  maxTokens: Int = 4096,
  temperature: Double = 0.7
)

trait Agent {
  def process(message: String)(implicit ec: ExecutionContext): Future[String]
  def getTools: List[String]
  def status: Status
}

trait Configurable {
  def configure(config: AgentConfig): Unit
}

abstract class BaseAgent(val name: String) extends Agent {
  protected var _status: Status = Idle
  override def status: Status = _status

  protected def validate(input: String): Boolean =
    input != null && input.trim.nonEmpty
}

class SynapseAgent(
  name: String,
  private var config: AgentConfig = AgentConfig()
) extends BaseAgent(name) with Configurable {

  override def process(message: String)(implicit ec: ExecutionContext): Future[String] = {
    _status = Active
    Future {
      val result = callModel(message)
      _status = Idle
      result
    }
  }

  override def getTools: List[String] = List("search", "read", "write")

  override def configure(config: AgentConfig): Unit = {
    this.config = config
  }

  private def callModel(message: String): String = {
    // TODO: implement actual model call
    s"Response to: $message"
  }
}

object SynapseAgent {
  def apply(name: String): SynapseAgent = new SynapseAgent(name)
}

// Verschachtelung fuer den findParentType-Fall: Entry steht DIREKT am Anfang von
// Registry, ohne schliessende Klammer dazwischen. Bis Parser-Version 3 lieferte der
// Eltern-Typ hier die AEUSSERE Deklaration (Registry) statt Entry. describe und
// describeTwice sowie lookup und count decken zusaetzlich den haeufigeren Fall ab:
// ab der ZWEITEN Methode einer Klasse ging der Eltern-Typ ganz verloren.
class Registry {
  class Entry(val key: String) {
    def describe(): String = key

    def describeTwice(): String = key + key
  }

  def lookup(key: String): Option[Entry] = None

  def count(): Int = 0
}

// FIXME: add proper error handling
